### ISSUE TITLE
Possible undocumented bug-fix/regression in 1.2.0 when using depends_on and modules with provider blocks

### ISSUE THREAD

**Issue by chrislwade** [hashicorp/terraform#31114]
### Terraform Version
```
Terraform v1.1.9
on darwin_amd64

Your version of Terraform is out of date! The latest version
is 1.2.0. You can update by downloading from https://www.terraform.io/downloads.html
```

and

```
Terraform v1.2.0
on darwin_amd64
```

### Terraform Configuration Files

See https://github.com/chrislwade/terraform-depends-on-module-with-provider.

### Expected Behavior

Either this example should continue to work (init, validate, and plan), or the release notes for 1.2.0 should be updated to reflect the fact that this scenario was fixed and not intended to work.

### Actual Behavior

Template fails init after updating to Terraform 1.2.0.

### Steps to Reproduce

1. Use Terraform version 1.1.9.
2. `terraform init`
3. `terraform validate`
4. Confirm template inits and validates.
5. Use Terraform version 1.2.0.
6. `terraform init`
7. Confirm template fails init.

### References

- #30639

**Comment by crw** [comment]
Thanks for the report!

**Comment by crw** [comment]
Duplicate of https://github.com/hashicorp/terraform/issues/31081 -- fix will be released in 1.2.1.
